src/Microsoft.Health.Fhir.SqlServer/Features/Search/SqlServerSearchService.cs

```diff
@@ -97,15 +98,58 @@ internal class SqlServerSearchService : SearchService
 
         public override async Task<SearchResult> SearchAsync(SearchOptions searchOptions, CancellationToken cancellationToken)
         {
-            SearchResult searchResult;
+            SqlSearchOptions sqlSearchOptions = new SqlSearchOptions(searchOptions);
+            SearchResult searchResult = await SearchImpl(sqlSearchOptions, SqlSearchType.Default, null, cancellationToken);
+            int resultCount = searchResult.Results.Count();
+            if (!sqlSearchOptions.IsSortWithFilter &&
+                searchResult.ContinuationToken == null &&
+                resultCount <= sqlSearchOptions.MaxItemCount &&
+                sqlSearchOptions.Sort != null &&
+                sqlSearchOptions.Sort.Count > 0 &&
+                sqlSearchOptions.Sort[0].searchParameterInfo.Code != KnownQueryParameterNames.LastUpdated)
+            {
+                // We seem to have run a sort which has returned less results than what max we can return.
+                // Let's determine whether we need to execute another query or not.
+                if ((sqlSearchOptions.Sort[0].sortOrder == SortOrder.Ascending && sqlSearchOptions.DidWeSearchForSortValue.HasValue && !sqlSearchOptions.DidWeSearchForSortValue.Value) ||
+                    (sqlSearchOptions.Sort[0].sortOrder == SortOrder.Descending && sqlSearchOptions.DidWeSearchForSortValue.HasValue && sqlSearchOptions.DidWeSearchForSortValue.Value))
+                {
+                    if (sqlSearchOptions.MaxItemCount - resultCount == 0)
+                    {
+                        // Since we are already returning MaxItemCount number of resources we don't want
+                        // to execute another search right now just to drop all the resources. We will return
+                        // a "special" ct so that we the subsequent request will be handled correctly.
```

Comment:
It's also possible here that there could be no more resources to actually search for. In this scenario a user would end up following the continuation token only to get 0 results.

Reply:
@brendankowitz Which situation would we prefer in this scenario?

- Have the user follow a continuation token and get 0 results
- Run an extra search to make sure there are more results

My preference would be to run the extra search. It is a little extra work on the server, but it saves the user a request/response cycle.